cmd/geth/verkle.go

```diff
@@ -0,0 +1,211 @@
+ `,
+			},
+			{
+				Name:      "dump",
+				Usage:     "Dump a verkle tree to a DOT file",
+				ArgsUsage: "<root> <key1> [<key 2> ...]",
+				Action:    expandVerkle,
+				Flags:     flags.Merge(utils.NetworkFlags, utils.DatabasePathFlags),
+				Description: `
+geth verkle dump <state-root> <key 1> [<key 2> ...]
+This command will produce a dot file representing the tree, rooted at <root>.
+in which key1, key2, ... are expanded.
+ `,
+			},
+		},
+	}
+)
+
+// recurse into each child to ensure they can be loaded from the db. The tree isn't rebuilt
+// (only its nodes are loaded) so there is no need to flush them, the garbage collector should
+// take care of that for us.
+func checkChildren(root verkle.VerkleNode, resolver verkle.NodeResolverFn) error {
+	switch node := root.(type) {
+	case *verkle.InternalNode:
+		for i, child := range node.Children() {
+			childC := child.ComputeCommitment().Bytes()
+
+			childS, err := resolver(childC[:])
+			if bytes.Equal(childC[:], zero[:]) {
+				continue
+			}
+			if err != nil {
+				return fmt.Errorf("could not find child %x in db: %w", childC, err)
+			}
+			// depth is set to 0, the tree isn't rebuilt so it's not a problem
+			childN, err := verkle.ParseNode(childS, 0, childC[:])
+			if err != nil {
+				return fmt.Errorf("decode error child %x in db: %w", child.ComputeCommitment().Bytes(), err)
+			}
+			if err := checkChildren(childN, resolver); err != nil {
+				return fmt.Errorf("%x%w", i, err) // write the path to the erroring node
+			}
+		}
+	case *verkle.LeafNode:
+		// sanity check: ensure at least one value is non-zero
+
+		for i := 0; i < verkle.NodeWidth; i++ {
+			if len(node.Value(i)) != 0 {
+				return nil
+			}
+		}
+		return fmt.Errorf("Both balance and nonce are 0")
+	case verkle.Empty:
+		// nothing to do
+	default:
+		return fmt.Errorf("unsupported type encountered %v", root)
+	}
+
+	return nil
+}
+
+func verifyVerkle(ctx *cli.Context) error {
+	stack, _ := makeConfigNode(ctx)
+	defer stack.Close()
+
+	chaindb := utils.MakeChainDatabase(ctx, stack, true)
+	headBlock := rawdb.ReadHeadBlock(chaindb)
+	if headBlock == nil {
+		log.Error("Failed to load head block")
+		return errors.New("no head block")
+	}
+	if ctx.NArg() > 1 {
+		log.Error("Too many arguments given")
+		return errors.New("too many arguments")
+	}
+	var (
+		rootC common.Hash
+		err   error
+	)
+	if ctx.NArg() == 1 {
+		rootC, err = parseRoot(ctx.Args().First())
+		if err != nil {
+			log.Error("Failed to resolve state root", "error", err)
+			return err
+		}
+		log.Info("Rebuilding the tree", "root", rootC)
+	} else {
+		rootC = headBlock.Root()
+		log.Info("Rebuilding the tree", "root", rootC, "number", headBlock.NumberU64())
+	}
+
+	serializedRoot, err := chaindb.Get(rootC[:])
+	if err != nil {
+		return err
+	}
+	root, err := verkle.ParseNode(serializedRoot, 0, rootC[:])
+	if err != nil {
+		return err
+	}
+
+	if err := checkChildren(root, chaindb.Get); err != nil {
```

Comment:
I don't get it. Does this assume that the chaindb is already migrated to verkle?

Reply:
yes, this is the verification tool, the goals are:
1. verify what Erigon has produced
2. until the inport from Erigon has been finalized, our conversion PRs can be rebased on top of this so the output can be verified.